hashes/src/internal_macros.rs

```diff
@@ -27,9 +27,9 @@ macro_rules! hash_trait_impls {
         $crate::impl_debug_only!(Hash, { $bits / 8 }, $reverse $(, $gen: $gent)*);
 
         impl<$($gen: $gent),*> $crate::GeneralHash for Hash<$($gen),*> {
-            type Engine = HashEngine;
+            type Engine = HashEngine<$($gen),*>;
 
-            fn from_engine(e: HashEngine) -> Hash<$($gen),*> { Self::from_engine(e) }
+            fn from_engine(e: HashEngine<$($gen),*>) -> Hash<$($gen),*> { Self::from_engine(e) }
```

Comment:
OT but this should've used `Self::Engine` in the first place.

Reply:
I'll do a followup PR where I go through these macros and clean them up. There's a few things like this which seem to just be accumulated cruft from extensions to the macros turning slightly-unidiomatic things into ugly ones.

Reply:
I ended up fixing this as part of this patch set since doing so made the final patch more simple.